### PR TITLE
Make Host directly mandatory

### LINKED ISSUE #722: Default value for Host
In determining the request target, the following text is included after failing to find an authority in configuration, the authority form target, or the Host header:

> Otherwise, the authority component is assigned the default name configured for the server and, if the connection's incoming TCP port number differs from the default port for the target URI's scheme, then a colon (":") and the incoming port number (in decimal form) are appended to the authority component.

This is a security risk because it means that servers might not agree with clients about the identity of the resource that the request is directed to.  This is especially bad as a TLS certificate is good for every port and often good for multiple names, which can mean a fairly large scope over which an attacker can exploit this confusion.

The origin form description doesn't use 2119 language, it says "A Host header field is also sent", citing [-semantics](https://httpwg.org/http-core/draft-ietf-httpbis-semantics-latest.html#rfc.section.7.2).  And that section is a bit weaselly in that regard, using a "SHOULD...unless".

> The target URI's authority information is critical for handling a request. A user agent SHOULD generate Host as the first field in the header section of a request unless it has already generated that information as an ":authority" pseudo-header field.

If this were a "MUST ... unless" in -semantics, this problem would not exist (at least in theory).

Just to confuse things, the absolute-form says [this](https://httpwg.org/http-core/draft-ietf-httpbis-messaging-latest.html#rfc.section.3.2.2.p.6), which might be read to imply that Host truly is mandatory:

> A client MUST send a Host header field in an HTTP/1.1 request even if the request-target is in the absolute-form

The asterisk-form is the only other option that naturally lacks an authority, and that says nothing, so that might be fixed up similarly.

### PR DESCRIPTION
**Pull request by martinthomson**
The normative language in -semantics wasn't direct enough about the
requirement for Host to be present in requests.  There was good language in
-messaging that I eventually found though, so there isn't any real problem
here, just an opportunity to improve.

I did this by taking the SHOULD...unless text and making that MUST unless,
splitting off the SHOULD that relates to putting Host as the first field line
of the header section.

Also, add a note in -messaging about supplying a default value for the
authority.

Closes #722.

### CODE REVIEW

**Review by reschke — COMMENTED** [review]
Not sure whether we are changing something normative here. Roy?

File: draft-ietf-httpbis-semantics-latest.xml [hunk 1]
```diff
@@ -2554,9 +2554,10 @@ Sun Nov  6 08:49:37 1994         ; ANSI C's asctime() format
 </sourcecode>
 <t>
    The target URI's authority information is critical for handling a
-   request. A user agent &SHOULD; generate Host as the first field in the
-   header section of a request unless it has already generated that information
-   as an ":authority" pseudo-header field.
+   request. A user agent &MUST; generate a Host header field unless it has
+   already generated that information as an ":authority" pseudo-header field.
+   A user agent &SHOULD; generate Host as the first field in the header
+   section of a request.
```
Comment (royfielding) [inline draft-ietf-httpbis-semantics-latest.xml:2560]:
```suggestion
   request. A user agent &MUST; generate a Host header field in a request
   unless it sends that information as an ":authority" pseudo-header field.
   A user agent that sends Host &SHOULD; send it as the first field in the
   header section of a request.
```

**Review by royfielding — CHANGES_REQUESTED** [review]
The changes to Messaging are inappropriate, or at least in the wrong sections. This is describing how to interpret a received message, not when Host is required. It isn't as simple as assuming that every server is configured for a single host.

The change to Semantics is okay with the suggestion applied.

**Comment by martinthomson** [comment]
Not sure that I follow your comment regarding -messaging.  I made two changes:

1. Add a condition to the text that allows filling in the host/authority from server configuration.  This says that only applies if the request version is HTTP/1.0, that's all.
2. Note that filling in the host based on server configuration creates an exposure to an attack.

Which of these is inappropriate?

**Comment by royfielding** [comment]
> Which of these is inappropriate?

They both are. Consider a server behind a gateway that is removing TLS for them. The TLS configuration and requirements have already been handled by the gateway. The server we are talking about here might be getting its external 
hostname via the local config, or it might be a fixed value, or it might be set by the gateway as a :authority or in Host, or it might be an entirely different field being set within a separate TLS connection specific to the gateway-origin connection.

It's a totally different concern than the requirements you are restating about TLS, which should already be handled elsewhere. If they are not being described in the right section, then we should fix that in the section about how to properly terminate the UA-to-server TLS connection.

**Comment by martinthomson** [comment]
But this has nothing to do with TLS.  It is about what the request target is.  If the host is receiving a request that uses HTTP/1.0, it is vulnerable to attacks.  Though the attacks in practice only really make sense when TLS is used (because TLS prevents modifying the requests, but not the destination host and port), the problem remains.  Unless you are assuming that the link between the gateway and origin server is secured against this class of attack (to which I suggest you are welcome to smile and nod at the note, which is just a note).

**Comment by wtarreau** [comment]
I was about to suggest something and found that it's already Martin's point 1, so I'm as well in favor of it: HTTP/1.1 must have a Host, 1.0 may not have one but in this case it's determined only by configuration (even if the configuration maps destination IP:ports to identity or host names). I'd be worried about dynamically setting a Host field matching the ip:port where the request landed, this would typically reveal private IP:ports after a TLS unwrapping gateway that could be reflected in server error responses or redirects.

**Comment by royfielding** [comment]
Note that, while HTTP/1.1 requires that Host be sent in a request, it does not require that the Host be non-empty. In many cases, servers are deployed as a hierarchy that supplies content regardless of the Host used for access. Whether or not that is secure for a given origin is entirely dependent on the nature of that origin, not on the nature of HTTP.

**Comment by wtarreau** [comment]
LGTM, thanks Roy!

Regarding the empty Host field, I've used it quite a few times to determine a site name from an IP address, because nowadays plenty of HTTP sites blindly redirect to HTTPS, and just sending a `GET /` with an empty host is often enough to see a hard-coded redirect to the real name :-)

File: draft-ietf-httpbis-messaging-latest.xml [hunk 2]
```diff
@@ -727,6 +728,14 @@ Host: www.example.org:8001
    scheme, "://", authority, and combined path and query component.
 </li>
 </ul>
+<aside>
+  <t>
+    &Note; Though absence of the <x:ref>Host</x:ref> header might be tolerated
+    by servers for compatibility with HTTP/1.0, supplying a default authority
+    makes it possible for an attacker to subvert the checks a server performs
+    on the target URI (see <xref target="Semantics" x:rel="#https.origin"/>).
```
[the remark this answers is not in the document]
Comment (martinthomson) [reply draft-ietf-httpbis-messaging-latest.xml:736]: I'm not sure that this correctly captures the intent.  You are saying that the security properties are uniquely known to the server (using its configuration, which might involve knowledge of the gateway involved).

The standard I'm looking for is that the client and server agree on what the value is.  This text doesn't quite guarantee that.  It just says that you end up with only one result if Host is omitted, which is almost the same, but not quite.

You say that the default needs to be consistent with the certificate (I read that as saying that the default needs to be listed in subjectAltName).  Then you say that it needs to be unique (which I read as saying that there is only one default for any given certificate).

The problem is that this assumes a single certificate being in play.  If you have multiple, which is common, then the default depends on which certificate is selected.

The reason I opened the original issue was to ensure that we acknowledged that assuming a default when Host is absent carries a risk.  My understanding is that this risk was why Host was made mandatory.  I'm surprised that you think that an empty value is useful.  This change makes it much less clear that this risk exists.

The goal of this seems to be to allow for deployments where a gateway communicates with an origin server that has a single identity.  In this configuration the gateway ensures that the server does not receive requests for the wrong name, AND the server doesn't pay attention to Host.  Maybe the servercan't know its own name and uses some sort of private key management for connections from the gateway.  Why not be clearer about that deployment arrangement and say "A server that can uniquely identify a Host identity from context MAY use that identity as a default without this risk."

**Comment by royfielding** [comment]
This was overtaken by #802

**Pull request closed without merging**